// griddle-react/dist/module/module.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.connect = exports.RowDefinition = exports.ColumnDefinition = exports.plugins = exports.utils = exports.settingsComponentObjects = exports.selectors = exports.constants = exports.components = exports.actions = undefined;

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

var _actions = require('./actions');

var actions = _interopRequireWildcard(_actions);

var _components = require('./components');

var _components2 = _interopRequireDefault(_components);

var _constants = require('./constants');

var constants = _interopRequireWildcard(_constants);

var _dataSelectors = require('./selectors/dataSelectors');

var selectors = _interopRequireWildcard(_dataSelectors);

var _settingsComponentObjects = require('./settingsComponentObjects');

var _settingsComponentObjects2 = _interopRequireDefault(_settingsComponentObjects);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

var _core = require('./core');

var _core2 = _interopRequireDefault(_core);

var _legacyStyle = require('./plugins/legacyStyle');

var _legacyStyle2 = _interopRequireDefault(_legacyStyle);

var _local = require('./plugins/local');

var _local2 = _interopRequireDefault(_local);

var _position = require('./plugins/position');

var _position2 = _interopRequireDefault(_position);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var plugins = {
  CorePlugin: _core2.default,
  LegacyStylePlugin: _legacyStyle2.default,
  LocalPlugin: _local2.default,
  PositionPlugin: _position2.default
};

var ColumnDefinition = _components2.default.ColumnDefinition;
var RowDefinition = _components2.default.RowDefinition;

var connect = _utils2.default.connect;

exports.default = _index2.default;
exports.actions = actions;
exports.components = _components2.default;
exports.constants = constants;
exports.selectors = selectors;
exports.settingsComponentObjects = _settingsComponentObjects2.default;
exports.utils = _utils2.default;
exports.plugins = plugins;
exports.ColumnDefinition = ColumnDefinition;
exports.RowDefinition = RowDefinition;
exports.connect = connect;
